README
NAME
    Net::Telnet::Gearman - interact with a Gearman server through its telnet
    interface

SYNOPSIS
        use Net::Telnet::Gearman;

        my $session = Net::Telnet::Gearman->new(
            Host => '127.0.0.1',
            Port => 4730,
        );

        my @workers   = $session->workers();
        my @functions = $session->status();
        my $version   = $session->version();
        my $result    = $session->maxqueue( reverse => 15 );

        $session->shutdown('graceful');

DESCRIPTION
    This is currently only tested with Gearman v0.10.

METHODS
  new
    This is the same as in Net::Telnet except for that there is called
    "$self->open()" for you.

  workers
    This sends back a list of all workers, their file descriptors, their
    IPs, their IDs, and a list of registered functions they can perform.

    See also: Net::Telnet::Gearman::Worker

  status
    This sends back a list of all registered functions. Next to each
    function is the number of jobs in the queue, the number of running jobs,
    and the number of capable workers.

    See also: Net::Telnet::Gearman::Function

  maxqueue
    This sets the maximum queue size for a function. If no size is given,
    the default is used. If the size is negative, then the queue is set to
    be unlimited. This sends back a single line with "OK".

    Arguments:

    *   Function name

    *   Maximum queue size (optional)

  shutdown
    Shutdown the server. If the optional "graceful" argument is used, close
    the listening socket and let all existing connections complete.

    Arguments:

    *   "graceful" (optional)

  version
    Send back the version of the server.

AUTHOR
    Johannes Plunien <plu@cpan.org>

COPYRIGHT AND LICENSE
    Copyright 2009 by Johannes Plunien

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

SEE ALSO
    *   Net::Telnet

    *   <http://gearman.org/index.php?id=protocol>

REPOSITORY
    <http://github.com/plu/net-telnet-gearman/>



Changes
Revision history for Net-Telnet-Gearman

0.04000     Sun Oct 21 2012
            - Bundle Module::Install 1.06
              See also: http://goo.gl/TXzhQ

0.03000     Thu Feb 16 2012
            - Skip function lines that do not return numbers for
              the busy, queue or running columns
            - Proxy arguments given to status and workers method
              to the Net::Telnet getline method

0.02000     Thu Mar 11 2010
            - Refactor ::Function and ::Worker to be testable
              plus add tests for it
            - Fix RT#55451

0.01000     Thu Nov 01 2009
            - Initial release.


lib/Net/Telnet/Gearman/Worker.pm
package Net::Telnet::Gearman::Worker;

use strict;
use warnings;
use base qw/Class::Accessor::Fast/;

__PACKAGE__->mk_accessors(qw/file_descriptor ip_address client_id functions/);

=head1 NAME

Net::Telnet::Gearman::Worker

=head1 SYNOPSIS

    use Net::Telnet::Gearman;
    
    my $session = Net::Telnet::Gearman->new(
        Host => '127.0.0.1',
        Port => 4730,
    );
    
    my @workers = $session->workers();
    
    print Dumper @workers
    
    # $VAR1 = bless(
    #     {
    #         'client_id'       => '-',
    #         'file_descriptor' => '1',
    #         'functions'       => [ 'resize_image' ],
    #         'ip_address'      => '127.0.0.1'
    #     },
    #     'Net::Telnet::Gearman::Worker'
    # );

=head1 METHODS

=head2 file_descriptor

Returns the file descriptor of this worker.

=head2 ip_address

Returns the ip address this worker is connected from.

=head2 client_id

Returns the client id of this worker.

=head2 functions

Returns an arrayref of functions the worker is registered for.

=head1 AUTHOR

Johannes Plunien E<lt>plu@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2009 by Johannes Plunien

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

=over 4 

=item * L<Net::Telnet::Gearman>

=back

=cut

1;

lib/Net/Telnet/Gearman/Function.pm
package Net::Telnet::Gearman::Function;

use strict;
use warnings;
use base qw/Class::Accessor::Fast/;

__PACKAGE__->mk_accessors(qw/name queue busy free running/);

sub parse_line {
    my ( $package, $line ) = @_;

    my ( $name, $queue, $busy, $running ) = split /[\s]+/, $line;

    my $free = $running - $busy;

    return $package->new(
        {
            name    => $name,
            queue   => $queue,
            busy    => $busy,
            free    => $free,
            running => $running,
        }
    );
}

=head1 NAME

Net::Telnet::Gearman::Function

=head1 SYNOPSIS

    use Net::Telnet::Gearman;
    
    my $session = Net::Telnet::Gearman->new(
        Host => '127.0.0.1',
        Port => 4730,
    );
    
    my @functions = $session->status();
    
    print Dumper @functions
    
    # $VAR1 = bless(
    #     {
    #         'busy'    => 10,
    #         'free'    => 0,
    #         'name'    => 'resize_image',
    #         'queue'   => 74,
    #         'running' => 10
    #     },
    #     'Net::Telnet::Gearman::Function'
    # );

=head1 METHODS

=head2 name

Returns the name of the function.

=head2 queue

Returns the amount of queued jobs for this function.

=head2 busy

Returns the amount of workers currently working on this function.

=head2 free

Returns the amount of free workers that can do this function.

=head2 running

Returns the amount of running workers registered for this function.

=head1 AUTHOR

Johannes Plunien E<lt>plu@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2009 by Johannes Plunien

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

=over 4 

=item * L<Net::Telnet::Gearman>

=back

=cut

1;


lib/Net/Telnet/Gearman.pm
package Net::Telnet::Gearman;

use strict;
use warnings;
use base qw/Net::Telnet/;
use Net::Telnet::Gearman::Worker;
use Net::Telnet::Gearman::Function;

our $VERSION = '0.04000';

=head1 NAME

Net::Telnet::Gearman - interact with a Gearman server through its telnet interface

=head1 SYNOPSIS

    use Net::Telnet::Gearman;

    my $session = Net::Telnet::Gearman->new(
        Host => '127.0.0.1',
        Port => 4730,
    );

    my @workers   = $session->workers();
    my @functions = $session->status();
    my $version   = $session->version();
    my $result    = $session->maxqueue( reverse => 15 );

    $session->shutdown('graceful');

=head1 DESCRIPTION

This is currently only tested with Gearman v0.10.

=head1 METHODS

=head2 new

This is the same as in L<Net::Telnet> except for that there is
called C<< $self->open() >> for you.

=cut

sub new {
    my $class = shift;

    # There's a new cmd_prompt in town.
    my $self = $class->SUPER::new(@_) or return;

    $self->open();

    return $self;
}

=head2 workers

This sends back a list of all workers, their file descriptors,
their IPs, their IDs, and a list of registered functions they can
perform.

See also: L<Net::Telnet::Gearman::Worker>

This method accepts any parameters the L<Net::Telnet> C<getline>
method does accept.

=cut

sub workers {
    my ($self, @args) = @_;

    $self->print('workers');

    my @workers = ();

    while ( my $line = $self->getline(@args) ) {
        last if $line eq ".\n";

        push @workers, Net::Telnet::Gearman::Worker->parse_line($line);
    }

    return wantarray ? @workers : \@workers;
}

=head2 status

This sends back a list of all registered functions. Next to
each function is the number of jobs in the queue, the number of
running jobs, and the number of capable workers.

See also: L<Net::Telnet::Gearman::Function>

This method accepts any parameters the L<Net::Telnet> C<getline>
method does accept.

=cut

sub status {
    my ($self, @args) = @_;

    $self->print('status');

    my @functions = ();

    while ( my $line = $self->getline(@args) ) {
        last if $line eq ".\n";

        next unless my $row = Net::Telnet::Gearman::Function->parse_line($line);

        push @functions, $row;
    }

    return wantarray ? @functions : \@functions;
}

=head2 maxqueue

This sets the maximum queue size for a function. If no size is
given, the default is used. If the size is negative, then the queue
is set to be unlimited. This sends back a single line with "OK".

Arguments:

=over 4

=item * Function name

=item * Maximum queue size (optional)

=back

=cut

sub maxqueue {
    my ( $self, @args ) = @_;
    unshift @args, 'maxqueue';
    return $self->_print_and_getline(@args);
}

=head2 shutdown

Shutdown the server. If the optional "graceful" argument is used,
close the listening socket and let all existing connections
complete.

Arguments:

=over 4

=item * "graceful" (optional)

=back

=cut

sub shutdown {
    my ( $self, $graceful ) = @_;
    my @args = ('shutdown');
    push @args, 'graceful' if $graceful;
    return $self->_print_and_getline(@args);
}

=head2 version

Send back the version of the server.

=cut

sub version {
    my ($self) = @_;
    return $self->_print_and_getline('version');
}

sub _print_and_getline {
    my ( $self, @args ) = @_;
    $self->print( join ' ', @args );
    my $line = $self->getline();
    chomp $line;
    return $line;
}

=head1 AUTHOR

Johannes Plunien E<lt>plu@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright 2009 by Johannes Plunien

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

=over 4 

=item * L<Net::Telnet>

=item * L<http://gearman.org/index.php?id=protocol>

=back

=head1 REPOSITORY

L<http://github.com/plu/net-telnet-gearman/>

=cut

1;
